trieste/acquisition/function.py

```diff
@@ -676,3 +676,245 @@ def batch_ei(at: TensorType) -> TensorType:
+        """
+        :param dataset: The data to use to build the acquisition function.
+        :param model: The model over the specified ``dataset``.
+        :param pending_points: Points already chosen to be in the current batch (of shape [M,D]),
+            where M is the number of pending points and D is the search space dimension.
+        :return: An acquisition function.
+        """
+
+
+class LocallyPenalizedExpectedImprovement(SingleModelGreedyAcquisitionBuilder):
+    r"""
+    Builder of the acquisition function maker for greedily collecting batches by local
+    penalization.  The resulting :const:`AcquisitionFunctionMaker` takes in a set of pending
+    points and returns the expected improvment acquisiton function penalized around those points.
+
+    Local penalization allows us to perform batch Bayesian optimization with a standard (non-batch)
+    acqusition function. By iteratively building a batch of points by maximizing an acquisition
+    function that is clipped around the values correponding to locations close to the already
+    chosen (pending) points, local penalization provides diverse batches of canidadate points.
+
+    Local penalization is applied to the acquisition function multiplicatively, however, to
+    improve numerical stability, we perfom additive penalization in a log space.
+
+    For our penalization function, we implement the soft penalization strategy
+    of :cite:`Gonzalez:2016`, where an estimate of the objective function's Lipschitz
+    constant is used to control the size of penalization.
+
+    The Lipschitz constant and additionl penalization parameters are estimated once
+    when preparing the :const:`AcquisitionFunctionMaker` and reused for each
+    individual :const:`AcquisitionFunction`) it makes.
+
+    """
+
+    def __init__(self, search_space: SearchSpace, grid_size: int = 500):
+        """
+        :param search_space: The global search space over which the optimisation is defined.
+        :param grid_size: Size of the grid over which the Lipschitz constant is estimated.
+        We recommend scaling this with search space dimension.
+        """
+        self._search_space = search_space
+        if grid_size <= 0:
+            raise ValueError(f"grid_size must be positive, got {grid_size}")
+        self._grid_size = grid_size
+
+        self._lipschitz_constant = None
+        self._eta = None
+
+    def prepare_acquisition_function(
+        self,
+        dataset: Dataset,
+        model: ProbabilisticModel,
+        pending_points: TensorType = None,
+    ) -> AcquisitionFunction:
+        """
+        :param dataset: The data from the observer.
+        :param model: The model over the specified ``dataset``.
+        :param pending_points: The points we penalize with respect to.
+        :return: The (log) expected improvement penalized with respect to the pending points.
+        This function will raise :exc:`ValueError` if its first call does not
```

Comment:
You should put exceptions as a separate entry rather than as part of the return (though you were missing the indentation to do that anyway, which is why there was an error).

```suggestion
        :raise ValueError: if the first call does not
```